File: src/main/java/io/vproxy/vfx/ui/button/ImageButton.java
package io.vproxy.vfx.ui.button;

import io.vproxy.vfx.control.click.ClickEventHandler;
import io.vproxy.vfx.manager.image.ImageManager;
import javafx.event.Event;
import javafx.event.EventHandler;
import javafx.scene.Cursor;
import javafx.scene.image.ImageView;

public class ImageButton extends ImageView {
    private final double w;
    private final double h;
    private EventHandler<Event> handler;

    public ImageButton(String prefix, String suffix) {
        setCursor(Cursor.HAND);

        var normalImage = ImageManager.get().load(prefix + "-normal" + "." + suffix);
        w = normalImage.getWidth();
        h = normalImage.getHeight();
        var hoverImage = ImageManager.get().load(prefix + "-hover" + "." + suffix);
        var downImage = ImageManager.get().load(prefix + "-down" + "." + suffix);
        setImage(normalImage);
        var clickHandler = new ClickEventHandler() {
            @Override
            protected void onMouseEntered() {
                setImage(hoverImage);
            }

            @Override
            protected void onMouseExited() {
                setImage(normalImage);
            }

            @Override
            protected void onMousePressed() {
                setImage(downImage);
            }

            @Override
            protected void onMouseReleased() {
                setImage(hoverImage);
            }

            @Override
            protected void onMouseClicked() {
                var handler = ImageButton.this.handler;
                if (handler == null) {
                    return;
                }
                handler.handle(null);
            }
        };
        setOnMouseEntered(clickHandler);
        setOnMouseExited(clickHandler);
        setOnMousePressed(clickHandler);
        setOnMouseReleased(clickHandler);
    }

    public void setOnAction(EventHandler<? extends Event> handler) {
        //noinspection unchecked
        this.handler = (EventHandler<Event>) handler;
    }

    public void setScale(double v) {
        setFitWidth(w * v);
        setFitHeight(h * v);
    }
}

File: src-ts/src/ui/button/ImageButton.ts
import { ImageView, FXImage } from '../../javafx/ImageView.js';
import { ClickEventHandler } from '../../control/click/ClickEventHandler.js';
import { ImageManager } from '../../manager/image/ImageManager.js';

export type ImageButtonHandler = (e: unknown) => void;

export class ImageButton extends ImageView {
  private readonly w: number;
  private readonly h: number;
  private handler: ImageButtonHandler | null = null;

  constructor(prefix: string, suffix: string) {
    super();
    this.setCursor('pointer');

    const normalImage = ImageManager.get().load(prefix + '-normal' + '.' + suffix);
    const normalImg: FXImage = normalImage ?? new FXImage('');
    this.w = normalImg.width;
    this.h = normalImg.height;
    const hoverImage = ImageManager.get().load(prefix + '-hover' + '.' + suffix) ?? normalImg;
    const downImage = ImageManager.get().load(prefix + '-down' + '.' + suffix) ?? normalImg;
    this.setImage(normalImg);

    const outer = this;
    const clickHandler = new (class extends ClickEventHandler {
      protected override onMouseEntered(): void {
        outer.setImage(hoverImage);
      }

      protected override onMouseExited(): void {
        outer.setImage(normalImg);
      }

      protected override onMousePressed(): void {
        outer.setImage(downImage);
      }

      protected override onMouseReleased(): void {
        outer.setImage(hoverImage);
      }

      protected override onMouseClicked(): void {
        const h = outer.handler;
        if (h === null) {
          return;
        }
        h(null);
      }
    })();
    this.setOnMouseEntered(clickHandler.handleEntered);
    this.setOnMouseExited(clickHandler.handleExited);
    this.setOnMousePressed(clickHandler.handlePressed);
    this.setOnMouseReleased(clickHandler.handleReleased);
  }

  setOnAction(handler: ImageButtonHandler | null): void {
    this.handler = handler;
  }

  setScale(v: number): void {
    this.setFitWidth(this.w * v);
    this.setFitHeight(this.h * v);
  }
}
